h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -600,9 +603,12 @@ public void onModelFailure(Model failedModel, Consumer<Object[]> withFailedModel
               hyperparamIndices[i] = _random.nextInt(_currentHyperParams.get(_currentHyperParamNames[i]).length);
             }
             // check for aliases and loop if we've visited this combo before
-          } while (_visitedPermutationHashes.contains(integerHash(_currentHyperParams, _currentHyperParamNames, hyperparamIndices, _currentSubspace)));
-          
-          return hyperparamIndices;
+           if (!_visitedPermutationHashes.contains(integerHash(_currentHyperParams, _currentHyperParamNames, hyperparamIndices, _currentSubspace)))
+            return hyperparamIndices;
+          }
+
+          _exhausted = true;
```

Comment:
so basically, you mark it as exhausted on the first collision, which, imo, is a very pessismistic way to mark the entire hyperspace as exhausted.

I'd rather skip the indices if there is a collision, count successive collisions and finally mark the hyperspace as exhausted if the successive collisions are above a certain threshold (e.g. 1/1000 of hyperspace size?).

Reply:
I think you misunderstood the code. This isn't `first collision => exhausted == True`. I don't check for collisions as it would be too expensive (I'd have to compare the parameter configurations not just their hashes to find out if it's collision or not).

The implementation is basically: if you can't find a new configuration within `maxHyperSpaceSize` steps say the space is `exhausted`. There can be multiple collisions and yet the space wouldn't be considered `exhausted` if you can find another new configuration within the `maxHyperSpaceSize` steps.

Reply:
oh right, my bad! As this loop is executed each time we need a new set of params, we might encounter previously used params, which is not a collision.

This looks good then, thanks.